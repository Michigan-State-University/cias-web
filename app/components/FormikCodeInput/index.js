/**
 *
 * FormikCodeInput
 *
 */

import React, { memo, useRef, createRef, useState } from 'react';
import PropTypes from 'prop-types';
import { useField } from 'formik';
import times from 'lodash/times';

import { Input } from 'components/Input';
import Text from 'components/Text';
import Row from 'components/Row';
import Column from 'components/Column';

import { ErrorText } from './styled';

const KEY_BACKSPACE = 8;
const numericRegex = /^[0-9]{1}$/;
const checkNumericInput = inputValue => !numericRegex.test(inputValue);

function FormikCodeInput({ formikKey, label, codeLength }) {
  const [field, meta, helpers] = useField(formikKey);
  const { value, onBlur, name } = field;
  const { error, touched } = meta;
  const hasError = touched && error;
  const { setValue, setTouched } = helpers;

  const [lastKey, setLastKey] = useState();
  const [code, setCode] = useState(
    value?.split('') ?? Array(codeLength).fill(),
  );
  const inputsRefs = useRef([]);

  if (inputsRefs.current.length !== codeLength) {
    inputsRefs.current = Array(codeLength)
      .fill()
      .map((_, i) => inputsRefs.current[i] || createRef());
  }

  const handleInput = e => {
    const {
      target: { value: inputValue },
    } = e;
    if (checkNumericInput(inputValue)) e.target.value = '';
  };

  const handleChange = (e, index) => {
    const tempCode = [...code];
    ({
      target: { value: tempCode[index] },
    } = e);
    setCode(tempCode);
    setValue(tempCode.join(''));
    setTouched(true);
  };

  const handleKeyUp = (e, index) => {
    e.preventDefault();
    const { keyCode: key } = e;
    const {
      target: { value: inputValue },
    } = e;
    if (key === KEY_BACKSPACE && lastKey === KEY_BACKSPACE) {
      if (index - 1 >= 0) {
        inputsRefs.current[index - 1].current.focus();
      } else {
        inputsRefs.current[index].current.blur();
      }
      return;
    }
    if (key === KEY_BACKSPACE) {
      setLastKey(KEY_BACKSPACE);
      return;
    }
    setLastKey('');
    if (checkNumericInput(inputValue)) return;
    if (index + 1 >= codeLength) {
      inputsRefs.current[index].current.blur();
      return;
    }
    inputsRefs.current[index + 1].current.select();
    inputsRefs.current[index + 1].current.focus();
  };

  return (
    <Column align="center">
      <Text mb={5} width="fit-content">
        {label}
      </Text>
      <Row justify="around" px={20} align="center">
        {times(codeLength).map(index => (
          <Input
            key={`code-input-${index}`}
            name={name}
            width={`calc(90% / ${codeLength})`}
            maxLength="1"
            size="1"
            onInput={handleInput}
            onKeyUp={e => handleKeyUp(e, index)}
            onChange={e => handleChange(e, index)}
            ref={inputsRefs.current[index]}
            textAlign="center"
            onBlur={onBlur}
          />
        ))}
      </Row>
      {hasError && <ErrorText>{error.toString()}</ErrorText>}
    </Column>
  );
}

FormikCodeInput.propTypes = {
  formikKey: PropTypes.string.isRequired,
  label: PropTypes.string,
  codeLength: PropTypes.number,
};

FormikCodeInput.defaultProps = { codeLength: 6 };

export default memo(FormikCodeInput);
