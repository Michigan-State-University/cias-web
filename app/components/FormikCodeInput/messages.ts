import { defineMessages } from 'react-intl';

export const scope = 'app.containers.FormikCodeInput';

export default defineMessages({
  codeInputLabel: {
    id: `${scope}.codeInputLabel`,
    defaultMessage: 'Enter digit',
  },
});
