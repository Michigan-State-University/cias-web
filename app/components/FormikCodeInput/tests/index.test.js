/**
 *
 * Tests for FormikCodeInput
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React, { Fragment } from 'react';
import { render } from '@testing-library/react';
import { Formik } from 'formik';

import FormikCodeInput from '../index';

describe('<FormikCodeInput />', () => {
  let props;
  beforeAll(() => {
    props = {
      formikKey: 'test',
      placeholder: 'test',
      type: 'text',
      label: 'Label',
    };
  });
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <Formik initialValues={{ test: '' }} onSubmit={jest.fn()}>
        {() => (
          <>
            <FormikCodeInput {...props} />
          </>
        )}
      </Formik>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const { container } = render(
      <Formik initialValues={{ test: '' }} onSubmit={jest.fn()}>
        {() => (
          <>
            <FormikCodeInput {...props} />
          </>
        )}
      </Formik>,
    );
    expect(container).toMatchSnapshot();
  });
});
