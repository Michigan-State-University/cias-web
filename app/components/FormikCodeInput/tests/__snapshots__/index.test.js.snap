// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormikCodeInput /> Should render and match the snapshot 1`] = `
.c5 {
  padding: 12px;
  border-style: solid;
  border-width: 1px;
  border-color: #E9F1FB;
  border-radius: 5px;
  width: 95%;
  color: #2F3850;
  text-align: center;
}

.c5:focus {
  box-shadow: none;
  outline: none;
  border-color: #107969;
}

.c5:disabled {
  color: #ABBBD1;
}

.c1 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
  margin-bottom: 5px;
  font-weight: 700;
  color: #2F3850;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

.c2 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
  margin-bottom: 5px;
  color: #2F3850;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

.c4 {
  font-family: DM Sans;
  font-size: 13px;
  line-height: 130%;
  color: #2F3850;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: auto;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  justify-items: space-around;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-right: 20px;
  padding-left: 20px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media only screen and (min-width:768px) {

}

@media only screen and (min-width:960px) {

}

@media only screen and (min-width:1170px) {

}

@media only screen and (min-width:1360px) {

}

<div>
  <div
    class="c0"
  >
    <p
      class="c1"
      font-weight="bold"
      width="fit-content"
    >
      Label
    </p>
    <p
      class="c2"
      id="code-input-general"
      width="fit-content"
    >
      Enter digit
    </p>
    <div
      class="c3"
    >
      <div
        class="c0"
      >
        <p
          class="c4"
          id="code-input-1"
        >
          1
        </p>
        <input
          aria-labelledby="code-input-general code-input-1"
          class="c5"
          maxlength="1"
          name="test"
          size="1"
          type="text"
          width="95%"
        />
      </div>
      <div
        class="c0"
      >
        <p
          class="c4"
          id="code-input-2"
        >
          2
        </p>
        <input
          aria-labelledby="code-input-general code-input-2"
          class="c5"
          maxlength="1"
          name="test"
          size="1"
          type="text"
          width="95%"
        />
      </div>
      <div
        class="c0"
      >
        <p
          class="c4"
          id="code-input-3"
        >
          3
        </p>
        <input
          aria-labelledby="code-input-general code-input-3"
          class="c5"
          maxlength="1"
          name="test"
          size="1"
          type="text"
          width="95%"
        />
      </div>
      <div
        class="c0"
      >
        <p
          class="c4"
          id="code-input-4"
        >
          4
        </p>
        <input
          aria-labelledby="code-input-general code-input-4"
          class="c5"
          maxlength="1"
          name="test"
          size="1"
          type="text"
          width="95%"
        />
      </div>
      <div
        class="c0"
      >
        <p
          class="c4"
          id="code-input-5"
        >
          5
        </p>
        <input
          aria-labelledby="code-input-general code-input-5"
          class="c5"
          maxlength="1"
          name="test"
          size="1"
          type="text"
          width="95%"
        />
      </div>
      <div
        class="c0"
      >
        <p
          class="c4"
          id="code-input-6"
        >
          6
        </p>
        <input
          aria-labelledby="code-input-general code-input-6"
          class="c5"
          maxlength="1"
          name="test"
          size="1"
          type="text"
          width="95%"
        />
      </div>
    </div>
  </div>
</div>
`;
